File: frontend/src/lib/marketplace.ts
import api from './api'

export interface Rating {
  id: string
  rating: number
  comment?: string
  ease_of_use?: number
  reliability?: number
  performance?: number
  support?: number
  created_at: string
}

export interface Review {
  id: string
  title: string
  content: string
  rating: number
  is_verified: boolean
  is_helpful: number
  created_at: string
}

export interface VendorStats {
  average_rating: number
  total_ratings: number
  total_reviews: number
  rating_distribution: Record<number, number>
}

export const marketplaceApi = {
  createRating: async (vendorId: string, agentId: string, rating: number, comment?: string) => {
    const response = await api.post('/marketplace/ratings', {
      vendor_id: vendorId,
      agent_id: agentId,
      rating,
      comment,
    })
    return response.data
  },
  
  createReview: async (vendorId: string, agentId: string, title: string, content: string, rating: number) => {
    const response = await api.post('/marketplace/reviews', {
      vendor_id: vendorId,
      agent_id: agentId,
      title,
      content,
      rating,
    })
    return response.data
  },
  
  getAgentRatings: async (agentId: string, page: number = 1, limit: number = 20) => {
    const response = await api.get(`/marketplace/agents/${agentId}/ratings`, {
      params: { page, limit },
    })
    return response.data
  },
  
  getAgentReviews: async (agentId: string, page: number = 1, limit: number = 20) => {
    const response = await api.get(`/marketplace/agents/${agentId}/reviews`, {
      params: { page, limit },
    })
    return response.data
  },
  
  getVendorStats: async (vendorId: string): Promise<VendorStats> => {
    const response = await api.get(`/marketplace/vendors/${vendorId}/stats`)
    return response.data
  },
}

File: backend/app/models/marketplace.py
"""
Marketplace models for vendor ratings and reviews
"""
from sqlalchemy import Column, String, DateTime, Integer, ForeignKey, Text, Numeric, Boolean
from sqlalchemy.dialects.postgresql import UUID
import uuid
from datetime import datetime
from app.core.database import Base
import enum


class RatingValue(int, enum.Enum):
    """Rating value (1-5 stars)"""
    ONE = 1
    TWO = 2
    THREE = 3
    FOUR = 4
    FIVE = 5


class VendorRating(Base):
    """Vendor rating by end users"""
    __tablename__ = "vendor_ratings"
    
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    vendor_id = Column(UUID(as_uuid=True), ForeignKey("vendors.id"), nullable=False, index=True)
    agent_id = Column(UUID(as_uuid=True), ForeignKey("agents.id"), nullable=False, index=True)
    user_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False, index=True)
    tenant_id = Column(UUID(as_uuid=True), nullable=True, index=True)
    
    # Rating details
    rating = Column(Integer, nullable=False)  # 1-5 stars
    comment = Column(Text, nullable=True)
    
    # Rating categories
    ease_of_use = Column(Integer, nullable=True)  # 1-5
    reliability = Column(Integer, nullable=True)  # 1-5
    performance = Column(Integer, nullable=True)  # 1-5
    support = Column(Integer, nullable=True)  # 1-5
    
    # Metadata
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    
    # Relationships
    # vendor = relationship("Vendor", back_populates="ratings")
    # agent = relationship("Agent", back_populates="ratings")
    # user = relationship("User", foreign_keys=[user_id])


class VendorReview(Base):
    """Detailed vendor review"""
    __tablename__ = "vendor_reviews"
    
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    vendor_id = Column(UUID(as_uuid=True), ForeignKey("vendors.id"), nullable=False, index=True)
    agent_id = Column(UUID(as_uuid=True), ForeignKey("agents.id"), nullable=False, index=True)
    user_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False, index=True)
    tenant_id = Column(UUID(as_uuid=True), nullable=True, index=True)
    
    # Review content
    title = Column(String(255), nullable=False)
    content = Column(Text, nullable=False)
    rating = Column(Integer, nullable=False)  # 1-5
    
    # Review status
    is_verified = Column(Boolean, default=False)  # Verified purchase/usage
    is_helpful = Column(Integer, default=0)  # Helpful votes count
    is_approved = Column(Boolean, default=True)  # Moderation status
    
    # Metadata
    created_at = Column(DateTime, default=datetime.utcnow, index=True)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    
    # Relationships
    # vendor = relationship("Vendor", back_populates="reviews")
    # agent = relationship("Agent", back_populates="reviews")
    # user = relationship("User", foreign_keys=[user_id])

